File: Leetcode/Binary_Search/069.Sqrt(x)/69. Sqrt(x).java
// Approach #2: Newton Solution

class Solution {
    public int mySqrt(int x) {
        long ans = 1;
     
        while (ans * ans <= x) {
            ++ans;
        }
        
        return (int) ans - 1;
    }
}

// Time Complexity: O(log(x))
// Space Complexity: O(1)

// Approach #3: Binary Search

class Solution {
    public int mySqrt(int x) {
        if (x == 0) return 0;
        int start = 1, end = x;
        while (start < end) { 
            int mid = start + (end - start) / 2;
            if (mid <= x / mid && (mid + 1) > x / (mid + 1))// Found the result
                return mid; 
            else if (mid > x / mid)// Keep checking the left part
                end = mid;
            else
                start = mid + 1;// Keep checking the right part
        }
        return start;
    }
}

// Time Complexity: O(log(n))
// Space Complexity: O(1)

File: Leetcode/Binary_Search/069.Sqrt(x)/69. Sqrt(x).cpp
// Approach #1: Binary Search

class Solution {
public:
    int mySqrt(int x) {
        if (x == 1)
        {
            return 1;
        }
        long long int i = 1, j = x / 2, mid;
        while (j > i)
        {
            mid = i + (j - i) / 2;
            long long int ans = mid * mid;
            if (ans == x)
            {
                return mid;
            }
            else if (ans > x)
            {
                j = mid - 1;
            }
            else
            {
                i = mid + 1;
            }
        }
        if (j * j > x)
        {
            return j - 1;
        }
        return j;
    }
};

// Time Complexity: O(log(n))
// Space Complexity: O(1)